### PR TITLE
Jkmarx/extend data set V2 api

### PR DESCRIPTION
**Pull request by jkmarx**
- Extend data set api v2 in support of future dashboard data set filtering

### CODE REVIEW

File: refinery/core/test_views.py [hunk 1]
```diff
@@ -0,0 +1,679 @@
+import json
```
Comment (jkmarx) [inline refinery/core/test_views.py]: refactored and move the views into separate file

File: refinery/core/test_views.py [hunk 2]
```diff
@@ -0,0 +1,679 @@
...
+        )
+        self.put_response = self.view(self.put_request)
+        self.options_request = self.factory.options(
+            self.url_root,
+            data=self.node_json,
+            format="json"
+        )
+        self.options_response = self.view(self.options_request)
+
+    def test_unallowed_http_verbs(self):
```
Comment (jkmarx) [inline refinery/core/test_views.py]: edited, removed get

File: refinery/core/test_views.py [hunk 3]
```diff
@@ -0,0 +1,679 @@
...
+        self.assertEqual(
+            self.put_response.data['detail'], 'Method "PUT" not allowed.')
+        self.assertEqual(
+            self.options_response.data['detail'], 'Method "OPTIONS" not '
+                                                  'allowed.')
+
+    def test_get_data_set_with_auth(self):
```
Comment (jkmarx) [inline refinery/core/test_views.py]: get unit tests

File: refinery/core/test_views.py [hunk 4]
```diff
@@ -0,0 +1,679 @@
...
+        patch_request = self.factory.patch(
+            urljoin(self.url_root, self.dataset.uuid),
+            {"title": new_title},
+        )
+        force_authenticate(patch_request, user=self.user)
+        patch_response = self.view(patch_request, self.dataset.uuid)
+        self.assertEqual(patch_response.status_code, 202)
+        self.assertEqual(patch_response.data.get('title'), new_title)
+
+    def test_is_filtered_data_set_returns_true_for_is_owner(self):
```
Comment (jkmarx) [inline refinery/core/test_views.py]: a bunch of unit test to check helper function used for filtering data set

File: refinery/ui/source/js/dashboard/controllers/data-sets-card-ctrl.js [hunk 5]
```diff
@@ -18,7 +18,7 @@
     '$uibModal',
     '$window',
     'DataSetSearchApi',
-    'dataSetService'
+    'dataSetV2Service'
```
Comment (jkmarx) [inline refinery/ui/source/js/dashboard/controllers/data-sets-card-ctrl.js]: update dashboard to use service

File: refinery/core/views.py [hunk 6]
```diff
@@ -697,7 +698,79 @@ class NodeViewSet(viewsets.ModelViewSet):
...
+            'is_public': params.get('public')
+        }
+        try:
+            filters['group'] = ExtendedGroup.objects.get(
+                id=params.get('group')
+            )
+        except Exception:
+            filters['group'] = None
+
+        query_data_sets = get_resources_for_user(
```
Comment (scottx611x) [inline refinery/core/views.py]: `data_sets_for_user` ?

File: refinery/core/views.py [hunk 7]
```diff
@@ -697,7 +698,79 @@ class NodeViewSet(viewsets.ModelViewSet):
...
+                        "DataSet with UUID: {} is invalid, and most likely is "
+                        "still being created".format(data_set.uuid)
+                    )
+                if self.is_filtered_data_set(data_set, filters):
+                    data_sets.append(data_set)
+        else:
+            data_sets = query_data_sets
+
+        serializer = DataSetSerializer(data_sets, many=True,
+                                       context={'request': request})
```
Comment (scottx611x) [inline refinery/core/views.py]: It seems odd to pass this into the serializer, I think I mentioned above that we could serialize a `DataSet's` `owner` and then check/add `is_owner` when we have access to `self.request.user` in the `get` method.

Alternatively, we could just pass the `User` in this context if that's what we truly need

File: refinery/core/views.py [hunk 8]
```diff
@@ -697,7 +698,79 @@ class NodeViewSet(viewsets.ModelViewSet):
...
+        params = request.query_params
+        filters = {
+            'is_owner': params.get('is_owner'),
+            'is_public': params.get('public')
+        }
+        try:
+            filters['group'] = ExtendedGroup.objects.get(
+                id=params.get('group')
+            )
+        except Exception:
```
Comment (scottx611x) [inline refinery/core/views.py]: Too broad of a clause here

File: refinery/core/serializers.py [hunk 9]
```diff
@@ -22,11 +22,26 @@ class DataSetSerializer(serializers.ModelSerializer):
             )]
     )
     description = serializers.CharField(max_length=5000)
+    is_owner = serializers.SerializerMethodField('check_is_owner')
+    public = serializers.SerializerMethodField('check_is_public')
```
[the remark this answers is not in the document]
Comment (jkmarx) [reply refinery/core/serializers.py]: I was trying to minimize the differences btwn the data set 1 api and data set 2 api response. Do we think it's worthwhile to go ahead and depart from it? @scottx611x

File: refinery/core/serializers.py [hunk 10]
```diff
@@ -22,11 +22,26 @@ class DataSetSerializer(serializers.ModelSerializer):
...
+
+    def check_is_owner(self, data_set):
```
[the remark this answers is not in the document]
Comment (jkmarx) [reply refinery/core/serializers.py]: sure

File: refinery/core/serializers.py [hunk 11]
```diff
@@ -22,11 +22,26 @@ class DataSetSerializer(serializers.ModelSerializer):
...
+
+    def check_is_public(self, data_set):
+        is_public = data_set.is_public()
+        return is_public
 
     class Meta:
         model = DataSet
         fields = ['title', 'accession', 'summary', 'description', 'slug',
-                  'uuid']
+                  'uuid', 'modification_date', 'id', 'is_owner', 'public']
```
[the remark this answers is not in the document]
Comment (jkmarx) [reply refinery/core/serializers.py]: Yes, the current launch pad uses data set ids for caching.

File: refinery/core/views.py [hunk 12]
```diff
@@ -697,7 +698,79 @@ class NodeViewSet(viewsets.ModelViewSet):
...
+        check_own = filter.get('is_owner')
+        owner = data_set.get_owner()
+        check_public = filter.get('is_public')
+        is_public = data_set.is_public()
+        group = filter.get('group')
+        group_perms = None
+        if group:
+            group_perms = get_perms(group, data_set)
+
+        if check_own and check_public and group:
```
[the remark this answers is not in the document]
Comment (jkmarx) [reply refinery/core/views.py]: We need to first check the params required and then return a response based on that. Ex: user wants a public & owned data set but with the logic above it'd return all the owned data sets for the filter.

File: refinery/core/serializers.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (scottx611x) [reply refinery/core/serializers.py]: Oh I didn't even realize, that would actually be more consistent!

File: refinery/core/test_views.py (hunk 4 above)
Comment (jkmarx) [reply refinery/core/test_views.py]: fixed by mocking the user object

File: refinery/core/views.py [hunk 13]
```diff
@@ -697,7 +698,79 @@ class NodeViewSet(viewsets.ModelViewSet):
...
+
+    def get(self, request):
+        params = request.query_params
+        filters = {
+            'is_owner': params.get('is_owner'),
+            'is_public': params.get('public')
+        }
+        try:
+            filters['group'] = ExtendedGroup.objects.get(
+                id=params.get('group')
```
[the remark this answers is not in the document]
Comment (jkmarx) [reply refinery/core/views.py]: yes, but trying to be consistent with current data set api v1

**Pull request merged**